### PR TITLE
refactor(test_utils): use json to run collection and add run time edit

### LINKED ISSUE #3806: [BUG] collections fail to import if variable is not a string
Newman runner i.e., newman-dir does support dynamic values but only string variables and not other variables say integer. Hence, variables like `"{{value}}"` is the only way for you to export the collection json file to its directory structure and having the variables like `{{value}}` will result in the `dir-export` / `dir-import` commands to fail.  

This bug affects collections that use dynamic values for amount fields. One such connector is NMI which has a hard limit on transactions i.e., once a payment for a n amount is done, the payment cannot be repeated up until the next 20 mins. If not, the payment is flagged as duplicate. The only work around for this is to use dynamic values. But that is restricted because of this bug with `newman-dir`. 

One work around that we can do is to:

- export the collection to its json format
- remove double quotes (`\"`) for integer fields
- run the collection -- json and not dir
- and all this should happen in run time

Example:

```sh
newman dir-import -o /tmp/generated_collection.json
perl -pi -e 's/\\"{{value}}\\"/{{value}}/g' /tmp/generated_collection.json
newman run /tmp/generated_collection.json
```

We can use `sed` / `regex` instead of `perl` as well.

Reference:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9a9eef47fb334cb686b2fff360e09536
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4e5ead4663c85da803b0a2c2030852eb
https://regex101.com/r/G6iapi/1
https://www.regular-expressions.info/named.html

### PR DESCRIPTION
**Pull request by pixincreate**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [x] CI/CD

## Description

This PR refactors test_utils crate's `newman_runner` to support dynamic values through postman variables.

The main issue with `newman-dir` is that in order to export the collection, the variables must be of type string (Example: Variable value must be of type `"{{value}}"` and not `{{value}}`) without which the user cannot import or export the collection in directory format which is a bummer.

With this refactor, we've introduced a validation to look after `connector_name` and if it is `nmi` or `powertranz`, we'll be doing the collection refactor in run time (collection refactor here means that before the collection is run, we'll be updating the json collection file by removing the double quotes in the integer values (`amount`) at least in 40+ places) which is later restored after the collection is run.

With this change, we'll again be using `run` command to run the collections instead of `dir-run` going forward.

This also added some complexity with how we're running tests for `custom_headers` and removed that part of code from all the existing collections.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

In order to run NMI collection which is updated here at #3805, this change is needed.

This PR should also close #3806.

## How did you test it?

In #3805, the tests ran without any issues. Adding the screenshot below:

image

Yet another validation: https://github.com/juspay/hyperswitch/actions/runs/8092287211/job/22112758729?pr=3807 (i know why it failed, outdated collection-json file which will be updated on next release after this pr is merged)

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/test_utils/src/newman_runner.rs [hunk 1]
```diff
@@ -72,22 +89,25 @@ pub fn generate_newman_command() -> ReturnArgs {
...
+    */
 
     let mut newman_command = Command::new("newman");
-    newman_command.args(["dir-run", &collection_path]);
+    newman_command.args(["run", &collection_path]);
     newman_command.args(["--env-var", &format!("admin_api_key={admin_api_key}")]);
     newman_command.args(["--env-var", &format!("baseUrl={base_url}")]);
 
-    if let Some(auth_type) = inner_map.get(&connector_name) {
+    // validation of connector is needed here as a work around to the limitation of newman-dir
```
Comment (SanchithHegde) [inline crates/test_utils/src/newman_runner.rs:109]: Can we not call it `newman-dir` but `newman fork` maybe?

File: crates/test_utils/src/newman_runner.rs [hunk 2]
```diff
@@ -208,3 +228,76 @@ pub fn generate_newman_command() -> ReturnArgs {
...
+// the corresponding collection is refactored in run time
+pub fn connector_validate(connector_name: &str) -> &str {
+    let refactorable_connector_names = ["nmi", "powertranz"];
+
+    if refactorable_connector_names.contains(&connector_name) {
+        return refactor_collection(connector_name).unwrap_or(connector_name);
+    }
+
+    connector_name
+}
```
Comment (SanchithHegde) [inline crates/test_utils/src/newman_runner.rs:242]: "Refactor" and "validate" are not the best words to be used here. Can we find something better, more specific to what we're doing?

File: crates/test_utils/src/newman_runner.rs [hunk 3]
```diff
@@ -208,3 +228,76 @@ pub fn generate_newman_command() -> ReturnArgs {
...
+
+/*
+Existing issue with newman-dir is that, it requires you to pass variables like `{{value}}` within
```
Comment (SanchithHegde) [inline crates/test_utils/src/newman_runner.rs:245]: Again, let's avoid calling it `newman-dir`.

File: crates/test_utils/src/newman_runner.rs [hunk 4]
```diff
@@ -72,22 +89,28 @@ pub fn generate_newman_command() -> ReturnArgs {
...
-    // been added to the postman collection, those conditions are set to true and collections that have
-    // variables set up for certificate, will consider those variables and will fail.
+    /*
+    Newman runner
+    Certificate keys are added through secrets in CI, so there's no need to explicitly pass it as arguments.
+    It can be overridden by explicitly passing cartificates as arguments.
+
+    If the collection takes in ceritificates (stripe collection for example) during MCA step,
+    then Stripe's certificates will be passed mandatorily (for now).
+    If any other connector requires certificates to be passed, that has to be passed explicitly for now.
```
Comment (SanchithHegde) [inline crates/test_utils/src/newman_runner.rs:104]:
```suggestion
    If the collection requires certificates (Stripe collection for example) during the merchant connector account create step,
    then Stripe's certificates will be passed implicitly (for now).
    If any other connector requires certificates to be passed, that has to be passed explicitly for now.
```

File: crates/test_utils/src/newman_runner.rs [hunk 5]
```diff
@@ -232,12 +232,12 @@ pub fn generate_newman_command() -> ReturnArgs {
     }
 }
 
-// If the connector name exists in refactorable_connector_names,
+// If the connector name exists in dynamic_amount_connectors,
 // the corresponding collection is refactored in run time to remove double quotes
```
Comment (SanchithHegde) [inline crates/test_utils/src/newman_runner.rs:236]:
```suggestion
// the corresponding collection is modified at runtime to remove double quotes
```

File: crates/test_utils/src/newman_runner.rs [hunk 6]
```diff
@@ -33,15 +40,23 @@ struct Args {
...
+    pub file_modified_flag: Vec<Flag>,
     pub collection_path: String,
 }
 
-// Just by the name of the connector, this function generates the name of the collection dir
-// Example: CONNECTOR_NAME="stripe" -> OUTPUT: postman/collection-dir/stripe
+pub struct Flag {
+    pub modified_status: bool,
+    pub file_to_restore: Option<String>,
+}
```
Comment (SanchithHegde) [inline crates/test_utils/src/newman_runner.rs:50]: Why not just collect list of file paths to be restored later?

File: crates/test_utils/src/newman_runner.rs [hunk 7]
```diff
@@ -187,24 +216,123 @@ pub fn generate_newman_command() -> ReturnArgs {
...
+pub fn check_for_custom_headers(headers: Option<Vec<String>>, path: &str) -> Option<String> {
+    if let Some(headers) = &headers {
         for header in headers {
             if let Some((key, value)) = header.split_once(':') {
                 let content_to_insert =
                     format!(r#"pm.request.headers.add({{key: "{key}", value: "{value}"}});"#);
-                if insert_content(&collection_path, &content_to_insert).is_ok() {
-                    modified = true;
-                }
+                let _ = insert_content(path, &content_to_insert).is_ok();
```
Comment (SanchithHegde) [inline crates/test_utils/src/newman_runner.rs:232]: Why do you need the `is_ok()`?

**Review by likhinbopanna — APPROVED** [review]

**Comment by likhinbopanna** [comment]
Review and approved for postman collections only

**Review by Narayanbhat166 — APPROVED** [review]

File: postman/collection-dir/nmi/Flow Testcases/Happy Cases/Scenario9-Update amount with automatic capture/Payments - Confirm/request.json [hunk 8]
```diff
@@ -29,12 +29,7 @@
       "key": "Accept",
       "value": "application/json"
     },
-    {
-      "key": "x-feature",
-      "value": "router-custom",
-      "type": "text",
-      "disabled": true
-    },
+    ,
```
Comment (SanchithHegde) [inline postman/collection-dir/nmi/Flow Testcases/Happy Cases/Scenario9-Update amount with automatic capture/Payments - Confirm/request.json:32]: We ended up leaving an extra comma behind here, ...

File: postman/collection-dir/nmi/Flow Testcases/Happy Cases/Scenario9a-Update amount with manual capture/Payments - Confirm/request.json [hunk 9]
```diff
@@ -29,12 +29,7 @@
       "key": "Accept",
       "value": "application/json"
     },
-    {
-      "key": "x-feature",
-      "value": "router-custom",
-      "type": "text",
-      "disabled": true
-    },
+    ,
```
Comment (SanchithHegde) [inline postman/collection-dir/nmi/Flow Testcases/Happy Cases/Scenario9a-Update amount with manual capture/Payments - Confirm/request.json:32]: ... here, ...

File: postman/collection-dir/stripe/Flow Testcases/Happy Cases/Scenario23- Update Amount/Payments - Confirm/request.json [hunk 10]
```diff
@@ -29,12 +29,7 @@
       "key": "Accept",
       "value": "application/json"
     },
-    {
-      "key": "x-feature",
-      "value": "router-custom",
-      "type": "text",
-      "disabled": true
-    },
+    ,
```
Comment (SanchithHegde) [inline postman/collection-dir/stripe/Flow Testcases/Happy Cases/Scenario23- Update Amount/Payments - Confirm/request.json:32]: ... and here.

Broke Postman tests for folks who had pulled the latest `main` branch.

**Pull request merged**